### PR TITLE
add partition endpoint for views along with validation on which

### PR DESCRIPTION
**Pull request by garrensmith**
Creating a new PR because the previous PR #1480 was closed and then I force pushed which means I cannot open it again.

This is a PR for the partitioned-wip branch. It adds some validation around partition work and a new url. The final url is still TBD but and can be changed in this PR easily at a later date.

### CODE REVIEW

File: src/chttpd/src/chttpd_db.erl [hunk 1]
```diff
@@ -250,20 +251,83 @@ handle_view_cleanup_req(Req, Db) ->
     ok = fabric:cleanup_index_files_all_nodes(Db),
     send_json(Req, 202, {[{ok, true}]}).
 
+
+handle_partition_design_req(#httpd{
+        path_parts=[DbName, <<"_partition">>, PartitionKey, _Design, Name, <<"_",_/binary>> = Action | _Rest]
```
Comment (rnewson) [inline src/chttpd/src/chttpd_db.erl]: In my work I'm calling this simply "Partition". For example, you might put documents in a partition called foo. It seems redundant to call that a partition key.

File: src/chttpd/src/chttpd_db.erl [hunk 2]
```diff
@@ -250,20 +251,83 @@ handle_view_cleanup_req(Req, Db) ->
...
+    }=Req, Db) -> 
+    validate_partition_key(PartitionKey),
+
+    case chttpd:qs_value(Req, "partition") of
+        undefined ->
+            ok;
+        _ ->
+            throw({
```
Comment (rnewson) [inline src/chttpd/src/chttpd_db.erl]: right idea, cwazy indentation.

File: src/chttpd/src/chttpd_db.erl [hunk 3]
```diff
@@ -250,20 +251,83 @@ handle_view_cleanup_req(Req, Db) ->
...
+                bad_request,
+                <<"Partition key is not allowed in the query string">>
+            })
+     end,
+
+    case mem3:is_partitioned(DbName) of
+        false -> throw(<<"Database is not partitioned">>);
```
Comment (rnewson) [inline src/chttpd/src/chttpd_db.erl]: {bad_request, ..}

File: src/chttpd/src/chttpd_db.erl [hunk 4]
```diff
@@ -250,20 +251,83 @@ handle_view_cleanup_req(Req, Db) ->
...
+    end;
+handle_partition_design_req(_Req, _Db) ->
+    throw({bad_request, <<"missing partition key">>}).
+
+
+validate_partition_key(<<"_", _/binary>>) ->
+    throw({bad_request, <<"partition keys must not begin with an underscore">>});
+validate_partition_key(<<" ">>) ->
+    throw({bad_request, <<"partition keys cannot begin with a space">>});
+validate_partition_key(_) ->
```
Comment (rnewson) [inline src/chttpd/src/chttpd_db.erl]: this is ad-hoc. there is a function to validate document ids inside couch already, we should call that. As per the updated spec, we should call it for the partition half _and_ the docid half, each of which has to obey the docid rules.

File: src/chttpd/src/chttpd_db.erl [hunk 5]
```diff
@@ -250,20 +251,83 @@ handle_view_cleanup_req(Req, Db) ->
...
+    ok.
+
+
+check_ddoc_can_use_partitionkey(#doc{body={DDocBody}}) ->
```
Comment (rnewson) [inline src/chttpd/src/chttpd_db.erl]: this needs to be deeper down as we will have the same question in other places. As a placeholder, fine.

File: src/chttpd/src/chttpd_view.erl [hunk 6]
```diff
@@ -101,6 +106,37 @@ handle_temp_view_req(Req, _Db) ->
...
+
+    Restrictions = [
+        {include_docs, Args#mrargs.include_docs, true},
+        {conflicts, Args#mrargs.conflicts, true},
+        {stable, Args#mrargs.stable, true}
+    ],
+    lists:foreach(fun ({Param, Field, Value}) ->
+        case Field =:= Value of
+            true ->
+                Msg = iolist_to_binary(io_lib:format("~s is not allowed for a partition query", [Param])),
```
Comment (rnewson) [inline src/chttpd/src/chttpd_view.erl]: no need to flatten the iolist, I think? and possibly as simple as `Msg = [Param, <<" is not allowed for a partition query">>]` if Param is a list or binary.

File: src/chttpd/src/chttpd_view.erl [hunk 7]
```diff
@@ -101,6 +106,37 @@ handle_temp_view_req(Req, _Db) ->
...
+                throw({
+                    bad_request,
```
Comment (rnewson) [inline src/chttpd/src/chttpd_view.erl]: needless indentation for something that fits comfortably on a single line

**Comment by garrensmith** [comment]
closing for https://github.com/apache/couchdb/pull/1577

**Pull request closed without merging**